File: samples/features/vector-search/vector-search-query-typescript/src/embed.ts
import { readFileSync, writeFileSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import {
  DefaultAzureCredential,
  getBearerTokenProvider,
} from "@azure/identity";
import { AzureOpenAI } from "openai";
import { loadConfig } from "./config.js";

// ---------------------------------------------------------------------------
// Resolve paths relative to this file (ESM equivalent of __dirname)
// ---------------------------------------------------------------------------
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const BATCH_SIZE = 20;

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------
async function main(): Promise<void> {
  // 1. Load config (SQL vars not required for embedding generation)
  const config = loadConfig(false);

  // 2. Read raw hotel data (no vectors)
  const inputPath = resolve(__dirname, "../../data/HotelsData.JSON");
  const outputPath = resolve(__dirname, "../../data/HotelsData_Vector.json");
  const hotels: Record<string, unknown>[] = JSON.parse(
    readFileSync(inputPath, "utf-8")
  );
  console.log(`Generating embeddings for ${hotels.length} hotels...`);

  // 3. Set up Azure OpenAI client with DefaultAzureCredential
  const credential = new DefaultAzureCredential();
  const azureADTokenProvider = getBearerTokenProvider(
    credential,
    "https://cognitiveservices.azure.com/.default"
  );
  const openaiClient = new AzureOpenAI({
    endpoint: config.azureOpenAiEndpoint,
    azureADTokenProvider,
    apiVersion: "2024-10-21",
    timeout: 30_000,
    maxRetries: 3,
  });

  // 4. Generate embeddings in batches
  const descriptions = hotels.map((h) => h["Description"] as string);
  const allEmbeddings: number[][] = [];

  for (let i = 0; i < descriptions.length; i += BATCH_SIZE) {
    const batch = descriptions.slice(i, i + BATCH_SIZE);
    const response = await openaiClient.embeddings.create({
      model: config.azureOpenAiEmbeddingDeployment,
      input: batch,
    });
    for (const item of response.data) {
      allEmbeddings.push(item.embedding);
    }
    console.log(
      `  Embedded ${Math.min(i + BATCH_SIZE, descriptions.length)}/${descriptions.length}`
    );
  }

  // 5. Attach vectors to hotel data and write output
  const hotelsWithVectors = hotels.map((hotel, idx) => ({
    ...hotel,
    DescriptionVector: allEmbeddings[idx],
  }));

  writeFileSync(outputPath, JSON.stringify(hotelsWithVectors, null, 2), "utf-8");
  console.log(`Done. Wrote HotelsData_Vector.json`);
}

main().catch((err) => {
  console.error("Error:", err);
  process.exit(1);
});

File: samples/features/vector-search/vector-search-query-python/src/embed.py
"""Generate precomputed embeddings for the hotels dataset (Python quickstart).

Reads data/HotelsData.JSON (no vectors), calls Azure OpenAI in batches to
generate a text-embedding-3-small embedding per hotel description, and
writes data/HotelsData_Vector.json — mirroring the TypeScript reference
sample's embed.ts. Run this only if you need to regenerate the shipped
vector data (for example, after changing the embedding model).
"""

from __future__ import annotations

import json
from pathlib import Path

from azure.identity import DefaultAzureCredential, get_bearer_token_provider
from dotenv import load_dotenv
from openai import AzureOpenAI

from config import load_config

BATCH_SIZE = 20


def main() -> None:
    load_dotenv()
    config = load_config(require_sql=False)

    base = Path(__file__).resolve().parent.parent.parent / "data"
    input_path = base / "HotelsData.JSON"
    output_path = base / "HotelsData_Vector.json"

    with input_path.open("r", encoding="utf-8") as f:
        hotels = json.load(f)
    print(f"Generating embeddings for {len(hotels)} hotels...")

    credential = DefaultAzureCredential()
    token_provider = get_bearer_token_provider(credential, "https://cognitiveservices.azure.com/.default")
    client = AzureOpenAI(
        azure_endpoint=config.azure_openai_endpoint,
        azure_ad_token_provider=token_provider,
        api_version="2024-10-21",
        timeout=30.0,
        max_retries=3,
    )

    descriptions = [hotel["Description"] for hotel in hotels]
    all_embeddings: list[list[float]] = []

    for start in range(0, len(descriptions), BATCH_SIZE):
        batch = descriptions[start : start + BATCH_SIZE]
        response = client.embeddings.create(model=config.azure_openai_embedding_deployment, input=batch)
        all_embeddings.extend(item.embedding for item in response.data)
        done = min(start + BATCH_SIZE, len(descriptions))
        print(f"  Embedded {done}/{len(descriptions)}")

    hotels_with_vectors = [
        {**hotel, "DescriptionVector": all_embeddings[idx]} for idx, hotel in enumerate(hotels)
    ]

    with output_path.open("w", encoding="utf-8") as f:
        json.dump(hotels_with_vectors, f, indent=2)
    print("Done. Wrote HotelsData_Vector.json")


if __name__ == "__main__":
    main()
